subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultResolverResults.java

```diff
@@ -16,145 +16,287 @@
+        return new DefaultResolverResults(
+            resolvedLocalComponentsResult,
+            resolutionResult,
+            visitedArtifacts,
+            resolvedConfiguration,
+            artifactResolveState,
+            resolveException
+        );
     }
 
-    private static boolean isNonFatalError(ResolveException failure) {
-        return failure.getCause() instanceof GraphValidationException;
+    @Override
+    public ResolverResults updateResolutionResult(Function<ResolutionResult, ResolutionResult> updater) {
+        return new DefaultResolverResults(
+            resolvedLocalComponentsResult,
+            updater.apply(resolutionResult),
+            visitedArtifacts,
+            resolvedConfiguration,
+            artifactResolveState,
+            failure
+        );
     }
 
-    @Override
-    public void artifactsResolved(ResolvedConfiguration resolvedConfiguration, VisitedArtifactSet visitedArtifacts) {
-        this.resolvedConfiguration = resolvedConfiguration;
-        this.visitedArtifacts = visitedArtifacts;
-        this.artifactResolveState = null;
+    /**
+     * Create a new result representing the result of resolving build dependencies.
+     */
+    public static ResolverResults buildDependenciesResolved(
+        ResolutionResult resolutionResult,
+        ResolvedLocalComponentsResult resolvedLocalComponentsResult,
+        VisitedArtifactSet visitedArtifacts
+    ) {
+        return new DefaultResolverResults(
+            resolvedLocalComponentsResult,
+            resolutionResult,
+            visitedArtifacts,
+            null,
+            null,
+            null
+        );
     }
 
-    @Override
-    public ResolveException consumeNonFatalFailure() {
-        try {
-            return nonFatalFailure;
-        } finally {
-            nonFatalFailure = null;
-        }
+    /**
+     * Create a new result representing the result of resolving the dependency graph.
+     */
+    public static ResolverResults graphResolved(
+        ResolutionResult resolutionResult,
+        ResolvedLocalComponentsResult resolvedLocalComponentsResult,
+        VisitedArtifactSet visitedArtifacts,
+        @Nullable ArtifactResolveState artifactResolveState
+    ) {
+        return new DefaultResolverResults(
+            resolvedLocalComponentsResult,
+            resolutionResult,
+            visitedArtifacts,
+            null,
+            artifactResolveState,
+            null
+        );
     }
 
-    @Override
-    public Throwable getFailure() {
-        if (fatalFailure != null) {
-            return fatalFailure;
-        }
-        if (nonFatalFailure != null) {
-            return nonFatalFailure;
-        }
-        return null;
+    /**
+     * Create a new result representing the result of resolving the artifacts.
+     */
+    public static ResolverResults artifactsResolved(ResolverResults graphResults, ResolvedConfiguration resolvedConfiguration, VisitedArtifactSet visitedArtifacts) {
```

Comment:
These are static factory methods which construct new `ResolverResults` instances. `withFailure` returns a copy of an existing result, with a failure attached. How do you think the static factory methods should be moved?

Reply:
Isn't that what this is doing, returning a modified version of an existing result (`graphResults`)?

Reply:
I see what you mean now. Only this method and the second `failed` one accept a `ResolverResults`, so I was a bit confused. 

IMO these are less so modifying the existing graph results, but rather creating a new artifact results state which includes some data from the graph results. Since these two instances represent separate states and since they should really be two separate types, I would prefer to go in the opposite direction, by keeping these as static factory methods but passing the relevant fields from `graphResults` individually instead of the entire instance.

Reply:
Yeah, I think that would be better.